Add support for composition given by the host view

tomalec opened the pull request:
https://github.com/Starcounter/starcounter-include/issues/102

tomalec committed:
Add support for composition given by the host view

https://github.com/Starcounter/starcounter-include/issues/102

warpech commented on starcounter-include.html:
```diff
@@ -96,6 +96,18 @@
+             * @readonly
+             * @type {DocumentFragment}
+             */
+            get hostComposition() {
```
This is a very expensive function. I think it should be a method, not a getter.

tomalec replied:
I just wanted to have it consistent with `.defaultComposition` and `.temporaryComposition`, but maybe we shouldn't have those as properties as well.

warpech reviewed (commented):
Before I continue with the review, I have a major question. Don't you think `starcounter-include` is already complex enough? Perhaps this is the time to actually remove the `declarative-shadow-dom` handling from it to the separate custom element.

Or would the performance penalty of that be too big?

warpech commented on README.md:
```diff
@@ -18,6 +18,7 @@ For more details see articles:
 
 - Fallback composition - a single `<style>:host{display: block;}</style><slot></slot>` to display all things from light DOM in case no composition was provided
 - Default composition - composition (or multiple concatenated compositions) found in the partial HTML view (`<template is="declarative-shadow-dom">` part)
+- Host composition - composition found in the host HTML view (content of `<starcounter-include ...><template is="declarative-shadow-dom" mode="host">`). It's used to enforce by host view a given composition for any kind of elements.
```
On the today's meeting with @tomalec, we decided to call this `presentation="parent"`

tomalec commented:
In short, the performance penalty is big, and the feature itself is not more complex than default composition. Also, sc-include already covers all other sources of presentation, why should we make an exception for this one?

tomalec committed:
Renamce host to parent composition,

remove getter for parentComposition,
as suggested at https://github.com/Starcounter/starcounter-include/pull/103#pullrequestreview-131136906

tomalec commented:
@warpech would you mind another round of review?

warpech commented on starcounter-include.html:
```diff
@@ -284,6 +285,17 @@
+                const parentCompositionTemplate = Array.from(this.children)
+                    .find((element)=>{
+                        return element.matches
+                            && element.matches('template[is="declarative-shadow-dom"][presentation="parent"]')
```
I like the code now!

My only doubt is that we agreed that the parent composition `declarative-shadow-dom` can be determined by heuristics of being a child of `starcounter-include`.

In your implementation though, it is determined by explicit attribute. Do you think this is unavoidable for future compatibility with flattened views?

I think it is avoidable. The "flattener" could insert `presentation="default"` attribute automatically for default presentations when it is flattening them in DOM. This way we would know which `declarative-shadow-doms` are parent and which not.

tomalec replied:
Right now, we can. For flattened we will need a way to differentiate them.

But, what is important now, if we would do as you suggested, in run-time after content is stamped app dev will see in his/her dev tools:

```html
<starcounter-include>
  #shadow-root //let's find out where it came from?
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom">...</template>
</starcounter-include>
```
That visuall look almost same if there is parent presentation or not.

While with `presentation="parent"`, he/she could easily find the actual source
```html
<starcounter-include>
  #shadow-root //let's find out where it came from?
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom" presentation="parent">...</template>
</starcounter-include>
```

tomalec replied:
>  think it is avoidable. The "flattener" could insert `presentation="default"` attribute automatically for default presentations when it is flattening them in DOM. 

I think it's confusing that `presentation="default" ` would be set explicitly and `presentation="parent"` would be the default

tomalec replied:
Also to do that, we would have to parse the HTML, and may not cover the cases when `<tempalte is="declaratice-shadow-dom">` is added dynamically by some script.

warpech replied:
If `presentation` attribute is not provided, we can set both `presentation="default"` and `presentation="parent"` implicitly by `s-i`, based on what `s-i` thinks it is.

So your example from https://github.com/Starcounter/starcounter-include/pull/103#discussion_r198749151 becomes:

```html
<starcounter-include>
  #shadow-root //let's find out where it came from?
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom" presentation="default">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom" presentation="default">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom" presentation="default">...</template>
  <hundrets>
  <of>
  <elements>
  <template is="declarative-shadow-dom" presentation="parent">...</template>
</starcounter-include>
```

even when `presentation` attr was not given in any of the source files.

tomalec committed:
Expose presentation type used & guessed by the element,

as suggested at https://github.com/Starcounter/starcounter-include/pull/103#discussion_r198758589

tomalec commented:
@warpech is it better now?

warpech commented on test/composition/default-declarative-shadow-dom.html:
```diff
@@ -185,6 +207,25 @@ <h2>Custom Shadow DOM alternative composition</h2>
 
     });
+    describe('starcounter-include when imports a template that contains <code>template is="declarative-shadow-dom"</code> with <code>presentation</code> attribute other than `author`, when imports a template', function () {
+        const REFERENCE_COMPOSITION = REFERENCE_HOST;
```
Please rename `REFERENCE_HOST` to `REFERENCE_PARENT` throughout the project so that we use consistent naming.

warpech reviewed (approved):
Yes, this is perfect! Just a small comment

tomalec committed:
Rename reference presentation to `REFERENCE_PARENT` in tests,

to keep naming consistent,
as suggested at https://github.com/Starcounter/starcounter-include/pull/103#discussion_r199906994
